Update UART driver with specs, documentation, and read/2

UncleGrumpy opened the pull request:
Updates the UART driver to include function specs and documentation.
Fixes possible concurrency problems, and insufficient sized memory checks.
Now cleans up and returns errors when initializing with badarg parameters.
Adds new `uart:read/2` with a timeout parameter.

Closes #1446

These changes are made under both the "Apache 2.0" and the "GNU Lesser General
Public License 2.1 or later" license terms (dual license).

SPDX-License-Identifier: Apache-2.0 OR LGPL-2.1-or-later

UncleGrumpy committed:
Update uart driver with specs, types, and function documentation

Add type definitions for input parameters.
Add spec for all exported functions.
Add documentation for functions.

Signed-off-by: Winford <winford@object.stream>

petermm reviewed (approved):
works, great! even when doing hundreds of uart:read per second!

pguyot commented on src/platforms/esp32/components/avm_builtins/uart_driver.c:
```diff
@@ -342,6 +362,15 @@ static void uart_driver_do_read(Context *ctx, GenMessage gen_message)
+            }
+            uint64_t timeout = ((uint64_t) term_to_int(timeover) * 1000u) + esp_timer_get_time();
+            uart_data->timeout = timeout;
+            xTaskCreate(vTaskReadTimeout, "timeout", 1000, uart_data, 1, NULL);
```
Since you're adding esp_timer, let's not just use it to get the time.

You could use esp_timer_create.

Alternatively, you could move the timer logic up to Erlang and change the API of the port.

pguyot commented on src/platforms/esp32/components/avm_builtins/uart_driver.c:
```diff
@@ -96,10 +98,27 @@ static const AtomStringIntPair cmd_table[] = {
+static void vTaskReadTimeout(void * pvParameters)
+{
+    struct UARTData *uart_data = (struct UARTData *) pvParameters;
+    while (esp_timer_get_time() < uart_data->timeout) {
```
Let's avoid this busy loop.

pguyot commented on src/platforms/esp32/components/avm_builtins/uart_driver.c:
```diff
@@ -96,10 +98,27 @@ static const AtomStringIntPair cmd_table[] = {
+    while (esp_timer_get_time() < uart_data->timeout) {
+        vTaskDelay(1);
+    }
+    uart_data->reader_process_pid = term_invalid_term();
```
We might have a concurrency issue here.

UncleGrumpy commented on libs/eavmlib/src/uart.erl:
```diff
@@ -19,24 +19,133 @@
+%% @end
+%%-----------------------------------------------------------------------------
+-spec close(Pid :: pid()) -> ok | {error, _Reason :: term()}.
+close(Pid) when is_pid(Pid) ->
```
I started to do this, but I thought this PR was better suited for release-0.6 branch, where theses are still PIDs… my plan was to fix this in the feature branch, where this is already a port.

pguyot replied:
Right, I thought is_port/1 existed on main already, but it doesn't seem to.

UncleGrumpy committed:
Fix concurrency issues and error handling in ESP32 UART driver

Fixes possible concurrency problems by using a mutex for access to `reader_process_pid` and
`reader_ref_ticks` in `uart_data`.

Avoid possible overflow of the queue by handling all mesages in the interrupt callback.

Now cleans up and returns errors, with improved log messages when initalization fails with a badarg
in the configuration parameters, rather than aborting.

Fixes incorrect pin integers being cast to terms.

Pin numbers are validated for the chip the VM is compiled for before use to mitigate VM crashes from
bad input parameters.

Signed-off-by: Winford <winford@object.stream>

UncleGrumpy committed:
Add read with timeout to uart driver

Adds a new uart:read/2 that takes a timeout parameter for reads. If no data is received during the
timeout period `timeout` is returned, and the listener is removed allowing for another read without
getting the `{error, ealready}` error tuple.

Closes #1446

Signed-off-by: Winford <winford@object.stream>

UncleGrumpy commented:
I pushed changes that eliminate the use of any timer or task in the C code, and instead added a `cancel_read` command, handling the timeout and sending the cancel command from erlang, following the suggestion @pguyot made.

II believe this is much more efficient, and less resource intensive than creating a timer.

pguyot commented on src/platforms/esp32/components/avm_builtins/uart_driver.c:
```diff
@@ -347,6 +349,19 @@ static void uart_driver_do_read(Context *ctx, GenMessage gen_message)
+{
+    struct UARTData *uart_data = ctx->platform_data;
+
+    uart_data->reader_process_pid = term_invalid_term();
```
`uart_interrupt_callback` can run on the other scheduler on dual core esp32s. So let's take advantage of this PR to protect uart_data accesses with a mutex or Atomic fields if it's sufficient.

UncleGrumpy replied:
I think making the two fields `uart_data->reader_process_pid`, and `uart_data->reader_ref_ticks` atomic should mitigate the problem, since these are both touched during read, cancel_read, and the interrupt callback.

pguyot reviewed (changes requested):
Thank you for this new version! Besides the little bug on timeout result, could you please fix potential concurrency issues?

petermm commented:
Works great still.

nitpick:
Should it only take a positive integer as timeout value? `pos_integer()`

Tried `:uart.read(port, 0)` but it would never get a read, of course also a nonsensical value..

UncleGrumpy commented on libs/eavmlib/src/uart.erl:
```diff
@@ -19,24 +19,141 @@
+    case port:call(Pid, read, Timeout) of
+        {error, timeout} ->
+            port:call(Pid, cancel_read),
+            timeout;
```
I was not convinced that the return should be {error, timeout} for read/2, since timeout is an expected result... we either want data or a timeout. I am not opposed to making this an error, but it seemed a bit wrong to me.

petermm replied:
true, I prefer the error tuple though.

having a single atom returns leads to more defensive code needed imho.  

eg. there is an additional return type, that I need to handle.

And it might even catch me by out be surprise, "shipped to prod" etc.

```elixir
    case :uart.read(port, 0) do
      {:ok, read} ->
       #
      {:error, error} ->
        # 
      other ->
        # 
    end
```

UncleGrumpy replied:
Sounds good to me, just looking for the return "of least surprise". I will update the return and documentation.

pguyot replied:
I was also reacting to the spec that doesn't allow `timeout`.

UncleGrumpy replied:
Clearly I was on the fence about the proper return there. lol. It's fixed now.

UncleGrumpy commented:
> Thank you for this new version! Besides the little bug on timeout result, could you please fix potential concurrency issues?

Yes. I see what you mean now, I thought your comment about concurrency was related to the previous implementation, but I see now that you were talking about pre-existing concurrency problems.

pguyot commented:
> Yes. I see what you mean now, I thought your comment about concurrency was related to the previous implementation, but I see now that you were talking about pre-existing concurrency problems.

Exactly. And issues could be amplified by the new code.

UncleGrumpy commented on src/platforms/esp32/components/avm_builtins/uart_driver.c:
```diff
@@ -62,8 +62,8 @@ struct UARTData
-    term reader_process_pid;
-    uint64_t reader_ref_ticks;
+    _Atomic term reader_process_pid;
+    _Atomic uint64_t reader_ref_ticks;
```
Now I have some doubts about this fix, it seems like the `reader_process_pid` and `reader_ref_ticks` could get out of sync, if a process were to change the `reader_process_id`, while another was waiting to change it, and the first process was interrupted by a higher priority task (before changing `reader_ref_ticks`), and the second then was allowed to change `reader_process_id` and then `reader_ref_ticks`, before the first process was scheduled to resume.

UncleGrumpy replied:
Or both?

UncleGrumpy commented:
This should be ready for another look. I made more bug fixes that should be checked.

arpunk commented on CHANGELOG.md:
```diff
@@ -33,6 +34,13 @@ certain VM instructions are used.
+
+### Changed
+
+- ESP32 UART driver no longer aborts because of barargs in configuration, instead rasing an error
```
I think you meant `badargs`.

arpunk commented on libs/eavmlib/src/uart.erl:
```diff
@@ -19,24 +19,141 @@
+% The peripheral `Name' may be one of: `"UART0"' | `"UART1"' | `"UART2"' | `<<"UART0">>' | `<<"UART1">>' | `<<"UART2">>'.
+
+-type uart_opts() :: [
+    {tx, _Tx_pin :: integer()}
+    | {rx, _Rx_pin :: integer()}
+    | {rts, _Rts_pin :: integer()}
+    | {cts, _Cts_pin :: integer()}
+    | {speed, _Speed :: pos_integer()}
+    | {data_bits, 5..8}
+    | {stop_bits, 1 | 2}
+    | {event_queue_len, _Qlen :: pos_integer()}
```
Is there a reason for the underscore prefix?

UncleGrumpy replied:
I thought I was getting some warnings either during compile, from dialyzer, or gradualizer, but I can't reproduce that now. I will remove them.

pguyot commented on src/platforms/esp32/components/avm_builtins/uart_driver.c:
```diff
@@ -186,18 +214,23 @@ Context *uart_driver_create_port(GlobalContext *global, term opts)
     term rx_pin = get_uart_pin_opt(opts, RX_ATOM);
     term rts_pin = get_uart_pin_opt(opts, RTS_ATOM);
     term cts_pin = get_uart_pin_opt(opts, CTS_ATOM);
+    if (UNLIKELY(tx_pin < -1 || rx_pin < -1 || rts_pin < -1 || cts_pin < -1)) {
```
`get_uart_pin_opt` returns an int but here it's cast to a term and compared with an int.
This is highly suspicious and dirty. We should instead distinguish `UART_PIN_NO_CHANGE` which is defined from esp-idf's uart.h from an error code, for example by using an int *ok or by returning a boolean.

UncleGrumpy replied:
Indeed! I overlooked that bug, these are ultimately used as int, they never should have been cast to a term.

pguyot commented on src/platforms/esp32/components/avm_builtins/uart_driver.c:
```diff
@@ -402,22 +463,27 @@ static void uart_driver_do_close(Context *ctx, GenMessage gen_message)
+        result = OK_ATOM;
+    }
+
+    if (UNLIKELY(memory_ensure_free_with_roots(ctx, TUPLE_SIZE(2) + REF_SIZE, 1, &ref, MEMORY_CAN_SHRINK) != MEMORY_GC_OK)) {
```
You don't need `REF_SIZE` here as `ref` is already allocated.

pguyot commented on src/platforms/esp32/components/avm_builtins/uart_driver.c:
```diff
@@ -387,9 +448,9 @@ static void uart_driver_do_write(Context *ctx, GenMessage gen_message)
 
     int local_pid = term_to_local_process_id(pid);
-    if (UNLIKELY(memory_ensure_free_with_roots(ctx, TUPLE_SIZE(2), 1, &ref, MEMORY_CAN_SHRINK) != MEMORY_GC_OK)) {
+    if (UNLIKELY(memory_ensure_free_with_roots(ctx, TUPLE_SIZE(2) + REF_SIZE, 1, &ref, MEMORY_CAN_SHRINK) != MEMORY_GC_OK)) {
```
You don't need `REF_SIZE` here as `ref` is already allocated.

UncleGrumpy replied:
Indeed. Once the hammer was in my hand I started to see too many nails. lol.

pguyot commented on src/platforms/esp32/components/avm_builtins/uart_driver.c:
```diff
@@ -342,11 +387,27 @@ static void uart_driver_do_read(Context *ctx, GenMessage gen_message)
+    term pid = gen_message.pid;
+    term ref = gen_message.ref;
+
+    if (UNLIKELY(memory_ensure_free_with_roots(ctx, TUPLE_SIZE(2) + REF_SIZE, 1, &ref, MEMORY_CAN_SHRINK) != MEMORY_GC_OK)) {
```
You don't need `REF_SIZE` here as `ref` is already allocated.

pguyot commented on src/platforms/esp32/components/avm_builtins/uart_driver.c:
```diff
@@ -326,9 +371,9 @@ static void uart_driver_do_read(Context *ctx, GenMessage gen_message)
     if (count > 0) {
         int bin_size = term_binary_heap_size(count);
-        if (UNLIKELY(memory_ensure_free_with_roots(ctx, bin_size + TUPLE_SIZE(2) * 2, 1, &ref, MEMORY_CAN_SHRINK) != MEMORY_GC_OK)) {
+        if (UNLIKELY(memory_ensure_free_with_roots(ctx, bin_size + TUPLE_SIZE(2) * 2 + REF_SIZE, 1, &ref, MEMORY_CAN_SHRINK) != MEMORY_GC_OK)) {
```
You don't need `REF_SIZE` here as `ref` is already allocated.

pguyot commented on src/platforms/esp32/components/avm_builtins/uart_driver.c:
```diff
@@ -309,13 +354,13 @@ static void uart_driver_do_read(Context *ctx, GenMessage gen_message)
 
     if (uart_data->reader_process_pid != term_invalid_term()) {
-        if (UNLIKELY(memory_ensure_free_with_roots(ctx, TUPLE_SIZE(2) * 2 , 1, &ref, MEMORY_CAN_SHRINK) != MEMORY_GC_OK)) {
+        if (UNLIKELY(memory_ensure_free_with_roots(ctx, TUPLE_SIZE(2) * 2 + REF_SIZE , 1, &ref, MEMORY_CAN_SHRINK) != MEMORY_GC_OK)) {
```
You don't need `REF_SIZE` here as `ref` is already allocated.

pguyot commented on CHANGELOG.md:
```diff
@@ -33,6 +34,13 @@ certain VM instructions are used.
 - Fix a memory corruption caused by `binary:split/2,3`
+- Add missing documentation and function specs for uart driver
+- Fixed possible concurrency problems in ESP32 UART driver
+- Fixed memory under allocation when creating ESP32 UART return messages
```
There were many issues in this driver, but I'm not sure it was missing `REF_SIZE` allocation.

pguyot commented on CHANGELOG.md:
```diff
@@ -33,6 +34,13 @@ certain VM instructions are used.
+
+### Changed
+
+- ESP32 UART driver no longer aborts because of badargs in configuration, instead rasing an error
```
s/rasing/raising/

pguyot commented on src/platforms/esp32/components/avm_builtins/uart_driver.c:
```diff
@@ -443,13 +509,12 @@ static NativeHandlerResult uart_driver_consume_mailbox(Context *ctx)
         if (is_closed) {
             if (UNLIKELY(memory_ensure_free(ctx, TUPLE_SIZE(2) * 2 + REF_SIZE) != MEMORY_GC_OK)) {
```
This could be changed to `memory_ensure_free_with_roots`, and include ref as a root, reduce allocation by REF_SIZE and avoid to convert ref back and forth to uint64_t.

UncleGrumpy commented on libs/eavmlib/src/uart.erl:
```diff
@@ -19,24 +19,141 @@
+%%-----------------------------------------------------------------------------
+-spec write(Pid :: pid(), Data :: iodata()) -> ok | {error, _Reason :: term()}.
+write(Pid, Data) when is_pid(Pid) ->
+    case is_iolist(Data) of
```
Because my linter was telling me that is_iolist() is illegal in a guard, is that a lie?

pguyot replied:
Sounds like a good reason :-) Arguably, both could be in case clause, but probably not worth the effort.

UncleGrumpy replied:
I did want both in a guard, for efficiency, but opted for this since that was not possible. It seems an extra case would add a bit more overhead, but the behavior would be more consistent. In this case won't the user get an undefined function error if something other than a pid is passed, and a badarg if the `Data` is invalid? In that case it probably is worth the effort to change both to a case.

pguyot replied:
The actual error isn't undef but function_clause but you're correct.

UncleGrumpy replied:
Fixed that, now both will throw a badarg.

UncleGrumpy commented on src/platforms/esp32/components/avm_builtins/uart_driver.c:
```diff
@@ -21,14 +21,22 @@
 #include <driver/uart.h>
 
 #include <esp_log.h>
+#include <esp_pthread.h>
```
No, my mistake. I forgot to remove that when I dropped the use of the task.
But this does bring to my attention also that the following two freertos included are not needed either ;-)

UncleGrumpy commented on src/platforms/esp32/components/avm_builtins/uart_driver.c:
```diff
@@ -284,15 +324,20 @@ Context *uart_driver_create_port(GlobalContext *global, term opts)
     uart_data->uart_num = uart_num;
     ctx->native_handler = uart_driver_consume_mailbox;
     ctx->platform_data = uart_data;
+#ifndef AVM_NO_SMP
```
Of course, I see now it would be dangling, since the driver can never be closed.

pguyot commented on src/platforms/esp32/components/avm_builtins/uart_driver.c:
```diff
@@ -182,22 +202,28 @@ Context *uart_driver_create_port(GlobalContext *global, term opts)
+    int rx_pin = get_uart_pin_opt(opts, RX_ATOM);
+    int rts_pin = get_uart_pin_opt(opts, RTS_ATOM);
+    int cts_pin = get_uart_pin_opt(opts, CTS_ATOM);
+    if (UNLIKELY(tx_pin < UART_PIN_NO_CHANGE || rx_pin < UART_PIN_NO_CHANGE
```
It's better, but code is still inferring that -2 is less than `UART_PIN_NO_CHANGE`, which is a little bit hacky or ugly.

The function should take an additional parameter as suggested (`int *ok`) or we should define a new constant for -2 (`UART_BADARG` ?) and a static assert that it's less than `UART_PIN_NO_CHANGE`.

UncleGrumpy replied:
I went with the static assert option, since we only need to check the value of the returned integers. Using the *ok pointer would have required 4 individual checks, rather than one after all pins have been evaluated.

pguyot reviewed (commented):
Do you want to take advantage of this PR to change the `if (xQueueRemove(...))` into a `while`? It would make queue overflow less likely.

UncleGrumpy commented:
> Do you want to take advantage of this PR to change the `if (xQueueRemove(...))` into a `while`? It would make queue overflow less likely.

I  @almost overlooked this. Great idea, I added that change.

robinchew commented:
httpd_example cannot load index.html if you open UART a certain way using this PR. 

This issue does **NOT** occur in commit 4d38f900 of this PR.

Find my modified http_example code here https://github.com/robinchew/atomvm_lib/commit/033a22696dbc822f01c24fd11e4e7a416ecf1cd8 which is essentially the following code which spawns a process that opens UART a certain way before `httpd:start`:
```erlang
    spawn(fun() ->
        Port = uart:open("UART2", [{rx, 16}, {tx, 17}]), % need to set rx and tx, and maybe UART2 as well
        io:format("port ~p~n", [Port]),
    end),
    
    httpd:start(....)
```

Trying to load index.html would result with:
```bash
~$ curl -v http://192.168.1.43:8080/index.html
*   Trying 192.168.1.43:8080...
* connect to 192.168.1.43 port 8080 from 192.168.1.4 port 37574 failed: Connection refused
* Failed to connect to 192.168.1.43 port 8080 after 306 ms: Couldn't connect to server
* Closing connection
curl: (7) Failed to connect to 192.168.1.43 port 8080 after 306 ms: Couldn't connect to server
```

UncleGrumpy commented:
> Do you want to take advantage of this PR to change the `if (xQueueRemove(...))` into a `while`? It would make queue overflow less likely.

I did really want to do this, but realized we can't without a lot more rewriting, or requiring the uart_driver sit behind a gen_server... as we have it now if there is no message to read immediately, the caller can drop into a receive and get the next message, at which point they would likely use a read to get the next incoming message. There is no way for the application to know that we are about to send another message from the queue after the first message is received.

UncleGrumpy commented:
I did add log errors for FIFO overflow, parity and frame errors, and a warning when the uart buffer is full. At least this way application developers will be aware when these events occur.

pguyot commented on src/platforms/esp32/components/avm_builtins/uart_driver.c:
```diff
@@ -86,23 +95,34 @@ enum uart_cmd
     struct UARTData *uart_data = GET_LIST_ENTRY(listener, struct UARTData, listener);
 
     uart_event_t event;
     if (xQueueReceive(uart_data->rxqueue, (void *) &event, (TickType_t) portMAX_DELAY)) {
```
Let's fix this: delay should be 0 here (this is what caused an issue with 5ac66d62f3fab4baa6d3568023e12eb72f6a20cb).

Indeed, listener is only called if data is available in the queue, so we never need to wait.

UncleGrumpy replied:
By switching back to an if the issue that we were discussing on Telegram is no longer a problem, but I agree this can and should be 0 here. The problem was when I tried to make this a while loop, which doesn't work for the reason I posted about it as a follow-up to your comment about using `while` instead.

UncleGrumpy commented:
> httpd_example cannot load index.html if you open UART a certain way using this PR.
> 
> This issue does **NOT** occur in commit [4d38f90](https://github.com/atomvm/AtomVM/commit/4d38f9008c1c89eac5afa445b648c0bc71ccb493) of this PR.
>... 

Thank you for testing and the follow up here. This was a mistake, I believe it should work much better now.

bettio commented on CHANGELOG.md:
```diff
@@ -50,7 +53,16 @@ bug when handling errors from BIFs used as NIFs (when called with `CALL_EXT` and
 - Fixed call to funs such as fun erlang:'not'/1, that make use of BIFs
 - Fixed potential crashes or memory leaks caused by a mistake in calculation of reference counts
 and a race condition in otp_socket code
+<<<<<<< HEAD
```
I think you committed some conflict markers. Can you cleanup this?
